### ISSUE TITLE
Rule proposal: Exhaust enum in switch-statement

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#281]
> I would like a rule that switch statements operating on an enum property should fully exhaust the enum, either by a case for all values or by a default case.

Relevant TSLint issue: https://github.com/palantir/tslint/issues/696

**Comment by mysticatea** [comment]
I will work on this.

**Comment by ypresto** [comment]
It'd better to support enum-like types such as  `'red' | 'green' | 'blue'`, if technically possible..!

**Comment by mysticatea** [comment]
Unfortunately, because my PC got broken, I need more time to finish.

**Comment by bradzacher** [comment]
Is it better to create a more general implementation for this?

Rather than an enum-specific rule, it would probably be good to just have a rule which forces a switch to handle all cases (i.e. exhaust until the default case is `never`).

**Comment by mohsen1** [comment]
@bradzacher cases can have similar types but different values 

```ts

let a = {}
let b = {}

function foo(arg: {}) {
    switch (arg) {
        case a: return 'A'
        case b: return 'B'
        default:
            console.log(arg)
    }
}

console.log(foo(a))
console.log(foo(b))
```

**Comment by zenVentzi** [comment]
Any updates on this?

**Comment by bradzacher** [comment]
nope!
as with any enhancement, happy to accept PRs.

**Comment by ekilah** [comment]
@zenVentzi in the mean time, I've adopted this work around in my project and it works quite nicely for most, but not all, cases: https://github.com/palantir/tslint/issues/696#issuecomment-468640023

**Comment by jacobwgillespie** [comment]
Some or all of this may have been implemented: https://github.com/typescript-eslint/typescript-eslint/pull/972
